src/modm/driver/inertial/bno055.hpp

```diff
@@ -0,0 +1,660 @@
+		RegisterMapAddressOutOfRange = 5,
+		RegisterMapWriteError = 6,
+		BNOLowPowerModeNotAvailableForSelectedOperationMode = 7,
+		AccelerometerPowerModeNotAvailable = 8,
+		FusionAlgorithmConfigurationError = 9,
+		SensorConfigurationError = 0xA,
+	};
+
+public:
+	struct modm_packed
+	Data
+	{
+		template < class I2cMaster >
+		friend class Bno055;
+
+		inline Vector3f acceleration()        const { return Vector3f(r.acceleration).scaled(1.f/16); }
+		inline Vector3f magnetometer()        const { return Vector3f(r.magnetometer).scaled(1.f/16); }
+		inline Vector3f gyroscope()           const { return Vector3f(r.gyroscope).scaled(1.f/16); }
+		inline float heading()                const { return r.heading / 16.f; }
+		inline float roll()                   const { return r.roll / 16.f; }
+		inline float pitch()                  const { return r.pitch / 16.f; }
+		inline Quaternion<float> quaternion() const { return Quaternion<float>(r.quaternion).scale(1l << 14); }
+		inline Vector3f linear_acceleration() const { return Vector3f(r.linear_acceleration).scaled(1.f/100); }
+		inline Vector3f gravity()             const { return Vector3f(r.gravity).scaled(1.f/100); }
+		inline int8_t temperature()           const { return r.temperature; }
+
+	private:
+		struct Raw {
+			int16_t acceleration[3];
+			int16_t magnetometer[3];
+			int16_t gyroscope[3];
+			int16_t heading;
+			int16_t roll;
+			int16_t pitch;
+			int16_t quaternion[4];
+			int16_t linear_acceleration[3];
+			int16_t gravity[3];
+			int8_t temperature;
+		} r;
+		static constexpr uint8_t size = sizeof(Raw);
+	};
+}; // struct bno055
+
+/**
+ * @ingroup modm_driver_bno055
+ * @author	Niklas Hauser
+ */
+template < class I2cMaster >
+class Bno055 : public bno055, public modm::I2cDevice<I2cMaster, 2>
+{
+public:
+	/// Constructor, requires a bno055::Data object.
+	/// For I2c this also sets the address to 0b110101 (alternative: 0x1C).
+	inline Bno055(Data &data, uint8_t address=addr()):
+		I2cDevice<I2cMaster,2>(address), data(data) {}
+
+	bool inline
+	configureBlocking(OperationMode mode=OperationMode::NDOF)
+	{
+		return RF_CALL_BLOCKING(configure(mode));
+	}
+
+	inline modm::ResumableResult<bool>
+	configure(OperationMode mode=OperationMode::NDOF)
+	{
+		return updateRegister(Register::OPR_MODE, uint8_t(mode));
+	}
+
+	inline modm::ResumableResult<bool>
+	readData()
+	{
+		const uint8_t reg = uint8_t(Register::ACCEL_DATA_X_LSB);
+		RF_BEGIN();
+
+		if ((reg ^ prev_reg) & 0x80) {
+			buffer[0] = uint8_t(Register::PAGE_ID);
+			buffer[1] = reg >> 7;
+			this->transaction.configureWrite(buffer, 2);
+			RF_CALL( this->runTransaction() );
+			prev_reg = reg;
+		}
+
+		buffer[0] = reg;
+		this->transaction.configureWriteRead(buffer, 1, (uint8_t*)&data, Data::size);
+		RF_END_RETURN_CALL( this->runTransaction() );
+	}
+
+	inline modm::ResumableResult<bool>
+	updateRegister(Register regi, uint8_t setMask, uint8_t clearMask = 0xff)
+	{
+		uint8_t reg = uint8_t(regi);
```

Comment:
This should be `const`. If it would be changed anywhere, the data flow would be become very fragile with the `RF` macros.

Reply:
This API isn’t done yet. But the datasheet is a bit terrible (typical for Bosch).